File: src/fr/adrienbrault/idea/symfony2plugin/config/xml/inspection/XmlServiceArgumentInspection.java
package fr.adrienbrault.idea.symfony2plugin.config.xml.inspection;

import com.intellij.codeInspection.LocalInspectionTool;
import com.intellij.codeInspection.ProblemHighlightType;
import com.intellij.codeInspection.ProblemsHolder;
import com.intellij.ide.highlighter.XmlFileType;
import com.intellij.patterns.PlatformPatterns;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiElementVisitor;
import com.intellij.psi.PsiFile;
import com.intellij.psi.xml.XmlTag;
import com.intellij.psi.xml.XmlTokenType;
import fr.adrienbrault.idea.symfony2plugin.Symfony2ProjectComponent;
import fr.adrienbrault.idea.symfony2plugin.action.ServiceActionUtil;
import fr.adrienbrault.idea.symfony2plugin.action.quickfix.AddServiceXmlArgumentLocalQuickFix;
import fr.adrienbrault.idea.symfony2plugin.stubs.ContainerCollectionResolver;
import fr.adrienbrault.idea.symfony2plugin.util.PsiElementUtils;
import org.jetbrains.annotations.NotNull;

import java.util.List;

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
public class XmlServiceArgumentInspection extends LocalInspectionTool {

    private ContainerCollectionResolver.LazyServiceCollector lazyServiceCollector;

    @NotNull
    @Override
    public PsiElementVisitor buildVisitor(final @NotNull ProblemsHolder holder, boolean isOnTheFly) {

        PsiFile psiFile = holder.getFile();
        if(psiFile.getFileType() != XmlFileType.INSTANCE || !Symfony2ProjectComponent.isEnabled(psiFile.getProject())) {
            return super.buildVisitor(holder, isOnTheFly);
        }

        for (XmlTag xmlTag : ServiceActionUtil.getXmlContainerServiceDefinition(psiFile)) {
            visitService(xmlTag, holder);
        }

        this.lazyServiceCollector = null;

        return super.buildVisitor(holder, isOnTheFly);
    }

    protected void visitService(XmlTag xmlTag, @NotNull ProblemsHolder holder) {

        if(!ServiceActionUtil.isValidXmlParameterInspectionService(xmlTag)) {
            return;
        }

        final List<String> args = ServiceActionUtil.getXmlMissingArgumentTypes(xmlTag, false, getLazyServiceCollector(xmlTag));
        if (args == null) {
            return;
        }

        PsiElement childrenOfType = PsiElementUtils.getChildrenOfType(xmlTag, PlatformPatterns.psiElement(XmlTokenType.XML_NAME));
        if(childrenOfType == null) {
            return;
        }

        holder.registerProblem(childrenOfType, "Missing Argument", ProblemHighlightType.GENERIC_ERROR_OR_WARNING, new AddServiceXmlArgumentLocalQuickFix(args));
    }

    private ContainerCollectionResolver.LazyServiceCollector getLazyServiceCollector(XmlTag xmlTag) {

        if(this.lazyServiceCollector != null) {
            return this.lazyServiceCollector;
        }

        return this.lazyServiceCollector = new ContainerCollectionResolver.LazyServiceCollector(xmlTag.getProject());
    }

}

File: src/main/kotlin/fr/adrienbrault/idea/symfony2plugin/config/xml/inspection/XmlServiceArgumentInspection.kt
package fr.adrienbrault.idea.symfony2plugin.config.xml.inspection

import com.intellij.codeInspection.LocalInspectionTool
import com.intellij.codeInspection.ProblemHighlightType
import com.intellij.codeInspection.ProblemsHolder
import com.intellij.openapi.util.NotNullLazyValue
import com.intellij.patterns.PlatformPatterns
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiElementVisitor
import com.intellij.psi.xml.XmlTag
import com.intellij.psi.xml.XmlTokenType
import fr.adrienbrault.idea.symfony2plugin.Symfony2ProjectComponent
import fr.adrienbrault.idea.symfony2plugin.action.ServiceActionUtil
import fr.adrienbrault.idea.symfony2plugin.action.quickfix.AddServiceXmlArgumentLocalQuickFix
import fr.adrienbrault.idea.symfony2plugin.stubs.ContainerCollectionResolver
import fr.adrienbrault.idea.symfony2plugin.util.PsiElementUtils

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
open class XmlServiceArgumentInspection : LocalInspectionTool() {
    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
        if (!Symfony2ProjectComponent.isEnabled(holder.project)) {
            return super.buildVisitor(holder, isOnTheFly)
        }

        return MyPsiElementVisitor(holder)
    }

    private class MyPsiElementVisitor(private val holder: ProblemsHolder) : PsiElementVisitor() {
        private var serviceCollector: NotNullLazyValue<ContainerCollectionResolver.LazyServiceCollector>? = null

        override fun visitElement(element: PsiElement) {
            if (element is XmlTag) {
                visitService(element, holder, createLazyServiceCollector())
            }

            super.visitElement(element)
        }

        private fun visitService(
            xmlTag: XmlTag,
            holder: ProblemsHolder,
            lazyServiceCollector: NotNullLazyValue<ContainerCollectionResolver.LazyServiceCollector>
        ) {
            if (!ServiceActionUtil.isValidXmlParameterInspectionService(xmlTag)) {
                return
            }

            val args = ServiceActionUtil.getXmlMissingArgumentTypes(xmlTag, false, lazyServiceCollector.get())
            if (args.isEmpty()) {
                return
            }

            val childrenOfType = PsiElementUtils.getChildrenOfType(xmlTag, PlatformPatterns.psiElement(XmlTokenType.XML_NAME))
                ?: return

            holder.registerProblem(
                childrenOfType,
                "Missing argument",
                ProblemHighlightType.GENERIC_ERROR_OR_WARNING,
                AddServiceXmlArgumentLocalQuickFix(args)
            )
        }

        private fun createLazyServiceCollector(): NotNullLazyValue<ContainerCollectionResolver.LazyServiceCollector> {
            if (serviceCollector == null) {
                serviceCollector = NotNullLazyValue.lazy { ContainerCollectionResolver.LazyServiceCollector(holder.project) }
            }

            return serviceCollector!!
        }
    }
}
